index.js

```diff
@@ -125,6 +140,463 @@
+    var tree = Node('CONSTRAINTS', '',
+                     chain(toPairs(data), TypeVarConstraints));
+    return pipe(tree, [
+      traverse(addSeparator(' => ')),
+      traverse(rootChildrenSeparator(',')),
+      traverse(addSpaces),
+      traverse(wrapRootChildren)
+    ]);
+  };
+
+  //  Args :: [ExpType] -> Node
+  var Args = function(data) {
+    var tree = Node('ARGS', '', map(data, Arg));
+    return pipe(tree, [
+      traverse(addSeparator(' -> ')),
+      traverse(rootChildrenSeparator(' ->')),
+      traverse(addSpaces),
+      traverse(wrapNestedContainers)
+    ]);
+  };
+
+  //  Ret :: ExpType -> Node
+  var Ret = function(data) {
+    var tree = Node('RET', '', [
+      Arg(data)
+    ]);
+    return pipe(tree, [
+      traverse(addSpaces),
+      traverse(wrapNestedContainers)
+    ]);
+  };
+
+  //  Arg :: ExpType -> Node
+  var Arg = function(arg) {
+    return Node(
+      arg.type,
+      arg.name ? stripNamespace(arg.name) : arg.toString(),
+      map(getArgChildren(arg), Arg));
+  };
+
+  //  TypeVarConstraints :: [[String, Constraint]] -> Node
+  var TypeVarConstraints = function(typeVarConstraints) {
+    var typeVar = typeVarConstraints[0];
+    var constraints = typeVarConstraints[1];
+    return map(constraints, Constraint(typeVar));
+  };
+
+  //  Constraint :: String -> Constraint -> Node
+  var Constraint = function(typeVarName) {
+    return function(constraint) {
+      return Node('CONSTRAINT', getConstraintText(constraint), [
+        Node('VARIABLE', typeVarName, [])
+      ]);
+    };
+  };
+
+  //  getArgChildren :: ExpType -> [ExpType]
+  var getArgChildren = function(arg) {
+    switch (arg.type) {
+      case 'UNARY':  return [arg.$1];
+      case 'BINARY': return [arg.$1, arg.$2];
+      default:       return [];
+    }
+  };
+
+  //  insert :: [Node] -> Node -> Node -> Integer -> Void
+  var insert = function(children, target, newChild, offset) {
+    for (var idx = 0; idx < children.length; idx += 1) {
+      if (children[idx] === target) {
+        children.splice(idx + offset, 0, newChild);
+        break;
+      }
+    }
+  };
+
+  //  insertAfter :: [Node] -> Node -> Node -> Void
+  var insertAfter = function(children, target, newChild) {
+    insert(children, target, newChild, 1);
+  };
+
+  //  insertBefore :: [Node] -> Node -> Node -> Void
+  var insertBefore = function(children, target, newChild) {
+    insert(children, target, newChild, 0);
+  };
+
+  //  setStart :: Integer -> Node -> Node -> Integer
+  var setStart = function(acc, node, parent) {
+    node.start = acc;
+    return acc + node.text.length;
+  };
+
+  //  setEnd :: Integer -> Node -> Node -> Integer
+  var setEnd = function(acc, node, parent) {
+    node.end = isLeaf(node) ? node.start + node.text.length : acc;
+    return node.end;
+  };
+
+  //  showSignature :: Node -> String
+  var showSignature = function(tree) {
+    return traverseWith(function(acc, t) { return acc.concat(t.text); },
```

Comment:
What's the type of `acc` here? If `String`, let's use `acc + t.text`. If `Array a`, let's use `acc.concat([t.text])`.